{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load modules\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Ontology import Ontology\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read Ontology  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done constructing ontology\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "## Read ontology files\n",
    "# -- First parameter is a 2-column table of (gene, term) annotations\n",
    "# -- Second parameter is a 2-column table of (child term, parent term) pairs\n",
    "# -- This example is the S. cerevisiae (budding yeast) Gene Ontology used in Yu et al. Translation of Genotype to Phenotype by a Hierarchy of Cell Subsystems. Cell Syst. 2016 Feb 24;2(2):77-88.\n",
    "ont = Ontology('example/yeast_child_2_parent.txt', 'example/yeast_gene_2_term.txt')\n",
    "\n",
    "### Propagate  gene-to-term annotations.\n",
    "# -- If gene g is annotated to term t, then we ensure that g is also annotated to all ancestral terms of t\n",
    "ont.propagate_annotations()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make Genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# A list of gene pairs that were knocked out\n",
    "double_knockouts = [\\\n",
    " ('YGR135W', 'YER065C'),\n",
    " ('YOR085W', 'YKL133C'),\n",
    " ('YPL017C', 'YOR380W'),\n",
    " ('YDL192W', 'YAL005C'),\n",
    " ('YJL214W', 'YGR203W'),\n",
    " ('YKL213C', 'YPR109W'),\n",
    " ('YKL213C', 'YDR458C'),\n",
    " ('YDR185C', 'YDL048C'),\n",
    " ('YPL213W', 'YJR127C'),\n",
    " ('YLR242C', 'YOL095C') \\\n",
    "                 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "### Convert the list of double knockouts into a strain-by-genes dataframe.\n",
    "# --Each row represents a different genetic strain's \"genotype\", and each column is a set of genes in the genome.\n",
    "# --Each value represent the functional status of a gene.\n",
    "#   In this example, if a gene is knocked out, then its functional status is -1 (complete loss-of-functoin), and otherwise it is '0' (wildtype or normal function).\n",
    "#   More generally, one can imagine setting a fractional value between -1 and 0 to indicate a partial loss-of-function, or a value above 0 to indicate a gain-of-function\n",
    "genotypes = pd.DataFrame(0, index=['genotype_%s' % i for i in range(len(double_knockouts))], columns=ont.genes, dtype=np.int8)\n",
    "for i, gene_pair in enumerate(double_knockouts):\n",
    "    genotypes.loc['genotype_%s' % i, gene_pair] = -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>15S_rRNA</th>\n",
       "      <th>21S_rRNA</th>\n",
       "      <th>AWA1</th>\n",
       "      <th>ENA6</th>\n",
       "      <th>ENS2</th>\n",
       "      <th>FPG1</th>\n",
       "      <th>HRA1</th>\n",
       "      <th>ICR1</th>\n",
       "      <th>KHR1</th>\n",
       "      <th>KHS1</th>\n",
       "      <th>...</th>\n",
       "      <th>tX(XXX)L</th>\n",
       "      <th>tY(GUA)D</th>\n",
       "      <th>tY(GUA)F1</th>\n",
       "      <th>tY(GUA)F2</th>\n",
       "      <th>tY(GUA)J1</th>\n",
       "      <th>tY(GUA)J2</th>\n",
       "      <th>tY(GUA)M1</th>\n",
       "      <th>tY(GUA)M2</th>\n",
       "      <th>tY(GUA)O</th>\n",
       "      <th>tY(GUA)Q</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>genotype_0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_7</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype_9</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 6046 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            15S_rRNA  21S_rRNA  AWA1  ENA6  ENS2  FPG1  HRA1  ICR1  KHR1  \\\n",
       "genotype_0         0         0     0     0     0     0     0     0     0   \n",
       "genotype_1         0         0     0     0     0     0     0     0     0   \n",
       "genotype_2         0         0     0     0     0     0     0     0     0   \n",
       "genotype_3         0         0     0     0     0     0     0     0     0   \n",
       "genotype_4         0         0     0     0     0     0     0     0     0   \n",
       "genotype_5         0         0     0     0     0     0     0     0     0   \n",
       "genotype_6         0         0     0     0     0     0     0     0     0   \n",
       "genotype_7         0         0     0     0     0     0     0     0     0   \n",
       "genotype_8         0         0     0     0     0     0     0     0     0   \n",
       "genotype_9         0         0     0     0     0     0     0     0     0   \n",
       "\n",
       "            KHS1    ...     tX(XXX)L  tY(GUA)D  tY(GUA)F1  tY(GUA)F2  \\\n",
       "genotype_0     0    ...            0         0          0          0   \n",
       "genotype_1     0    ...            0         0          0          0   \n",
       "genotype_2     0    ...            0         0          0          0   \n",
       "genotype_3     0    ...            0         0          0          0   \n",
       "genotype_4     0    ...            0         0          0          0   \n",
       "genotype_5     0    ...            0         0          0          0   \n",
       "genotype_6     0    ...            0         0          0          0   \n",
       "genotype_7     0    ...            0         0          0          0   \n",
       "genotype_8     0    ...            0         0          0          0   \n",
       "genotype_9     0    ...            0         0          0          0   \n",
       "\n",
       "            tY(GUA)J1  tY(GUA)J2  tY(GUA)M1  tY(GUA)M2  tY(GUA)O  tY(GUA)Q  \n",
       "genotype_0          0          0          0          0         0         0  \n",
       "genotype_1          0          0          0          0         0         0  \n",
       "genotype_2          0          0          0          0         0         0  \n",
       "genotype_3          0          0          0          0         0         0  \n",
       "genotype_4          0          0          0          0         0         0  \n",
       "genotype_5          0          0          0          0         0         0  \n",
       "genotype_6          0          0          0          0         0         0  \n",
       "genotype_7          0          0          0          0         0         0  \n",
       "genotype_8          0          0          0          0         0         0  \n",
       "genotype_9          0          0          0          0         0         0  \n",
       "\n",
       "[10 rows x 6046 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# For intuition, show the genotypes dataframe\n",
    "display(genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genotype_0   -2\n",
       "genotype_1   -2\n",
       "genotype_2   -2\n",
       "genotype_3   -2\n",
       "genotype_4   -2\n",
       "genotype_5   -2\n",
       "genotype_6   -2\n",
       "genotype_7   -2\n",
       "genotype_8   -2\n",
       "genotype_9   -2\n",
       "dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# As a sanity check, note that the row sums of the dataframe are all -2 because the genotypes are double knockouts\n",
    "display(genotypes.sum(1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compute ontotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing 4698 terms with no mutations among samples\n"
     ]
    }
   ],
   "source": [
    "## Calculate a strain-by-term dataframe\n",
    "# -- Each row represents a genetic strain's ontotype, and each column represents a term in the ontology.\n",
    "# -- Each value is calculated by simply summing the functional status of genes in that term.\n",
    "#    In this example, a term's value is 0 if none of its gene have been knocked out, is -1 if exactly one gene has been knocked out, and -2 if two genes have been knocked out\n",
    "# -- This simple calculation was used in Yu et al. Translation of Genotype to Phenotype by a Hierarchy of Cell Subsystems. Cell Syst. 2016 Feb 24;2(2):77-88.\n",
    "#    However, one can imagine different formulas for calculating the ontotype.\n",
    "ontotypes = ont.get_features(genotypes, prop='matrix_mult', format=pd.DataFrame)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GO_0000002</th>\n",
       "      <th>GO_0000003</th>\n",
       "      <th>GO_0000060</th>\n",
       "      <th>GO_0000070</th>\n",
       "      <th>GO_0000087</th>\n",
       "      <th>GO_0000139</th>\n",
       "      <th>GO_0000166</th>\n",
       "      <th>GO_0000226</th>\n",
       "      <th>GO_0000267</th>\n",
       "      <th>GO_0000278</th>\n",
       "      <th>...</th>\n",
       "      <th>GO_0071842</th>\n",
       "      <th>GO_0071844</th>\n",
       "      <th>GO_0071944</th>\n",
       "      <th>GO_0072594</th>\n",
       "      <th>GO_0072655</th>\n",
       "      <th>GO_0080090</th>\n",
       "      <th>GO_0080129</th>\n",
       "      <th>GO_0090304</th>\n",
       "      <th>GO_00SUPER</th>\n",
       "      <th>GO_2000112</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>-2</td>\n",
       "      <td>-2</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 427 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   GO_0000002  GO_0000003  GO_0000060  GO_0000070  GO_0000087  GO_0000139  \\\n",
       "0           0           0           0           0           0           0   \n",
       "1           0           0           0           0           0           0   \n",
       "2           0          -1           0          -1          -1           0   \n",
       "3           0           0          -1           0           0           0   \n",
       "4           0           0           0           0           0           0   \n",
       "5           0           0           0           0           0           0   \n",
       "6           0           0           0           0           0           0   \n",
       "7           0           0           0           0           0           0   \n",
       "8           0           0           0           0           0           0   \n",
       "9          -1          -1           0           0           0          -1   \n",
       "\n",
       "   GO_0000166  GO_0000226  GO_0000267  GO_0000278     ...      GO_0071842  \\\n",
       "0           0           0           0           0     ...              -1   \n",
       "1           0           0           0           0     ...               0   \n",
       "2           0          -1           0          -1     ...              -1   \n",
       "3          -2           0          -1           0     ...              -1   \n",
       "4           0           0           0           0     ...               0   \n",
       "5           0           0           0           0     ...               0   \n",
       "6           0           0           0           0     ...               0   \n",
       "7           0           0           0           0     ...               0   \n",
       "8           0           0           0           0     ...               0   \n",
       "9          -1           0           0           0     ...              -1   \n",
       "\n",
       "   GO_0071844  GO_0071944  GO_0072594  GO_0072655  GO_0080090  GO_0080129  \\\n",
       "0          -1           0           0           0           0          -1   \n",
       "1           0           0           0           0           0           0   \n",
       "2           0           0           0           0          -1           0   \n",
       "3           0          -1          -1          -1           0           0   \n",
       "4           0          -1           0           0           0           0   \n",
       "5           0           0           0           0           0           0   \n",
       "6           0           0           0           0           0           0   \n",
       "7           0           0           0           0           0           0   \n",
       "8           0           0           0           0          -1           0   \n",
       "9           0           0           0           0           0           0   \n",
       "\n",
       "   GO_0090304  GO_00SUPER  GO_2000112  \n",
       "0           0          -2           0  \n",
       "1           0          -2           0  \n",
       "2          -2          -2          -1  \n",
       "3           0          -2           0  \n",
       "4           0          -2           0  \n",
       "5          -1          -2           0  \n",
       "6          -1          -2           0  \n",
       "7           0          -2           0  \n",
       "8          -2          -2          -1  \n",
       "9          -1          -2           0  \n",
       "\n",
       "[10 rows x 427 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# For intuition, show the ontotypes dataframe\n",
    "display(ontotypes)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
